lib/logstash/outputs/s3/file_repository.rb

```diff
@@ -96,9 +99,12 @@ def size
         end
 
         def remove_stale(k, v)
-          if v.stale?
-            @prefixed_factories.delete_pair(k, v)
-            v.delete!
+          stale_time = v.stale_time
+          v.with_lock do |factory|
+            if factory.current.size == 0 && (Time.now - factory.current.ctime) > stale_time
```

Comment:
This change orphans the `FileRepository::PrefixedValue#stale?` method, which is unit-tested, and brings the logic into a place where it cannot be unit tested.

The following keeps the unit-tested code in place, but requires `FileRepository::PrefixedValue#with_lock` to be backed by a reentrant lock (e.g., `Monitor` instead of `Mutex`):

```suggestion
          v.with_lock do |factory|
            if v.stale?
```